### PR TITLE
Book 3 Chapter 1-3 First Draft with Figures

### PR DESCRIPTION
**Pull request by trevordblack**
It's in. Holy moly it's in.

Took me a while, but I'm finally done with the first draft of half the book.

### CODE REVIEW

File: src/InOneWeekend/main.cc [hunk 1]
```diff
@@ -75,7 +75,7 @@ void random_spheres(scene& scene_desc) {
 }
 
 
-void main() {
+int main() {
```
Comment (hollasch) [inline src/InOneWeekend/main.cc:78]: Why change this back to an (ignored) `int` return type?

**Comment by trevordblack** [comment]
Clang6 on my ubuntu was throwing out errors for main that don't return an int. That version of clang is just more sensitive than most compilers. It's a small thing, but it was an error for me, I assumed it would be an error for a small slice of the audience

File: books/RayTracingTheRestOfYourLife.html [hunk 2]
```diff
@@ -423,184 +571,504 @@
...
+
+   ![Figure [uniform-dist]: A uniform distribution](../images/fig-3.05-uniform-dist.jpg)
+
+Our, new, crude approximation for $\frac{r}{2}$ is nonuniform (but only just):
+
+    ![Figure [approx-f]: A nonuniform distribution for r/2](../images/fig-3.06-nonuniform-dist.jpg)
+
+We had the analytical solution to the integration above, so we could very very easily solve for the
+50% value. We could also solve for this experimentally. If we had a function that couldn't solve
+for the integration for--or didn't _want_ to solve the integration for--we could get an experimental
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:833]: Reword; I can't parse this.

File: books/RayTracingTheRestOfYourLife.html [hunk 3]
```diff
@@ -423,184 +571,504 @@
...
+result close to the real value. Let's take the function:
+
+    $$ p(x) = e^{\frac{-x}{2 \pi}} sin^2(x) $$
+
+Which looks a little something like this:
+
+    ![Figure [exp-sin2]: A function that we don't want to solve analytically](../images/fig-3.07-exp-sin2.jpg)
+
+From $0$ to $2\pi$ as example:
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:842]: Jumping from the text right into the full code is quite abrupt here. It would help to spend more time here to set up the problem and approach that the following code tackles, so the reader has a chance to reflect on how _they_ might solve this problem.

It takes some code reading and parsing to work out that the following code is creating an ordered list of x, f(x), and then walks the result until the accumulated sum is about halfway.

In fact, I think it would be more performant and more clear to just fill an array of {x, f(x)}, and then call `std::sort` (introduced in the BVT code) before searching for the halfway mark.

File: books/RayTracingTheRestOfYourLife.html [hunk 4]
```diff
@@ -423,184 +571,504 @@
...
+        std::cout << "Average = " << sum / N << '\n';
+        std::cout << "Area under curve = " << 2 * pi * sum / N << '\n';
+        std::cout << "Halfway = " << halfway_point << '\n';
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [est-halfway]: <kbd>[estimate_halfway.cc]</kbd> Estimating the 50% point of a function]
+
+This code snippet isn't too different from what we had before. We're still solving for the sum
+over an interval (0 to $2\pi$). Only this time, we're also storing and sorting all of our samples by
+their input and output. We then use this to determine how many of the original samples sum up to
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:898]: “We then use this to determine the point at which they subtotal up to half of the sum ...”

File: books/RayTracingTheRestOfYourLife.html [hunk 5]
```diff
@@ -423,184 +571,504 @@
...
+half of the sum across the total interval. Once we know that $j$ of our first samples sum up to half
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:899]: “... that our first _j_ samples sum up to ...”

File: books/RayTracingTheRestOfYourLife.html [hunk 6]
```diff
@@ -423,184 +571,504 @@
...
+of the total sum, we know that the $j\text{th}$ $x$ roughly cooresponds to our halfway point:
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:900]: corresponds

File: books/RayTracingTheRestOfYourLife.html [hunk 7]
```diff
@@ -423,184 +571,504 @@
...
+distributions of nonuniform numbers. This divide and conquer method of producing nonuniform
+distributions is used somewhat commonly in practice, although there are much more efficient means
+of doing so than a simple binary partition. If you have an arbitrary function that you wish to use
+as the PDF for a distribution, you'll want to research the _Metropolis-Hastings Algorithm_.
+
+Approximating Distributions
+---------------------------
+
+This was a lot of math and work to build up a couple of notions. Let's return to our initial PDF and
+enumerate a new term. Where we don't explicitly set a probability, we assume to be zero. So for our
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:933]: “Let's return to our initial PDF ~and enumerate a new term~.”

Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:933]: “We assume the PDF to be zero where we don't explicitly set a probability.”

File: books/RayTracingTheRestOfYourLife.html [hunk 8]
```diff
@@ -423,184 +571,504 @@
...
+example from the beginning of the chapter, $p(r) = 0$, for $r < 0$ and $r > 2$. _i.e._, the
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:934]: Start sentences with “That is, ...” instead of a Latin abbreviation.

File: books/RayTracingTheRestOfYourLife.html [hunk 9]
```diff
@@ -423,184 +571,504 @@
...
+probability of generating a random number there is zero. We can rewrite our $p(r)$ in piecewise
+fashion:
+
+    $$ p(r)=\begin{cases}0 & r < 0 \\ \frac{r}{2} & 0 < r < 2 \\ 0 & r > 2  \\  \end{cases} $$
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:938]: $$ p(r)=\begin{cases}0 & r < 0 \\ \frac{r}{2} & 0 \leq r \leq 2 \\ 0 & r > 2  \\  \end{cases} $$

→ 0 ≤ r ≤ 2

File: books/RayTracingTheRestOfYourLife.html [hunk 10]
```diff
@@ -423,184 +571,504 @@
...
+    $$ P(x) =  \int_{-\infty}^{x}  p(x') dx' $$
+
+Or,
+
+    $$ P(x) = \text{area}(p(x'), -\infty, x) $$
+
+Which is the amount of _cumulative_ probability from $-\infty$. If we take the integration outlined
+above, we get the piecewise $P(r)$:
+
+    $$ P(r)=\begin{cases}0 & r < 0 \\ \frac{r^2}{4} & 0 < r < 2 \\ 1 & r > 2  \\  \end{cases} $$
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:958]: 0 ≤ r ≤ 2

File: books/RayTracingTheRestOfYourLife.html [hunk 11]
```diff
@@ -423,184 +571,504 @@
...
-want a function `f()` that when we call it as `f(random_double())` we get a return value with a PDF
-$\frac{x^2}{4}$. We don’t know what that is, but we do know that 25% of what it returns should be
-less than 1.0, and 75% should be above 1.0. If $f()$ is increasing, then we would expect $f(0.25) =
-1.0$. This can be generalized to figure out $f()$ for every possible input:
+If we evaluate the CDF, $P(r)$, at $r = 1.0$, we get:
 
-  $$ f(P(x)) = x $$
+    $$ P(1.0) = \frac{1}{4} $$
+
+This says _a random variable plucked from our PDF has a 25% change of being 1 or lower_. We want a
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:974]: chance

File: books/RayTracingTheRestOfYourLife.html [hunk 12]
```diff
@@ -423,184 +571,504 @@
...
 <div class='together'>
+Looking closer at the equality:
+
+    $$ f(P(x)) = x $$
+
+That just means that $f()$ just undoes whatever $P()$ does. So,
+
+    $$ f(d) = P^{-1}(x) $$
+
 The -1 means “inverse function”. Ugly notation, but standard. For our purposes, if we have PDF $p()$
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1024]: I'd assume this is commonly known. If you still wish to keep it, perhaps make it a parenthetical comment.

File: books/RayTracingTheRestOfYourLife.html [hunk 13]
```diff
@@ -705,15 +1182,14 @@
 Let’s review now because that was most of the concepts that underlie MC ray tracers.
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:1182]: I know it's not part of your changes, but how about rewording this to “That was most of the concepts that underlie Monte Carlo ray tracers. Let's review:”

File: books/RayTracingTheRestOfYourLife.html [hunk 14]
```diff
@@ -705,15 +1182,14 @@
...
-  2. You pick a PDF $p$ that is non-zero over $[a,b]$
+  2. You pick a PDF $p$ that is non-zero and non-negative over $[a,b]$
   3. You average a whole ton of $\frac{f(r)}{p(r)}$ where $r$ is a random number with PDF $p$.
 
 Any choice of PDF $p$ will always converge to the right answer, but the closer that $p$
 approximates $f$, the faster that it will converge.
 </div>
 
 
-
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:716]: We've been using three blank lines for H1 separation. Two lines for H2.

**Review by hollasch — APPROVED** [review]
_Phew!_ Nice work.

File: src/InOneWeekend/main.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (trevordblack) [reply src/InOneWeekend/main.cc:78]: Commented above, copied here:

Clang6 on my ubuntu was throwing out errors for main that don't return an int. That version of clang is just more sensitive than most compilers. It's a small thing, but it was an error for me, I assumed it would be an error for a small slice of the audience

I don't care either way.

File: books/RayTracingTheRestOfYourLife.html [hunk 15]
```diff
@@ -382,30 +382,178 @@
...
+
+If we take the number of samples $N$ and take the limit as $N$ goes to $\infty$, then we get the
+following:
+
+    $$ E[f(x') | a \leq x' \leq b]  = \lim_{N \to +\infty} \frac{1}{N} \sum_{i=0}^{N-1} f(x_i) $$
+
+Within the continuous interval $[a,b]$, the expected value of continuous function $f(x')$ can be
+perfectly represented by summing an infinite number of random points within the interval. As this
+number of points approaches $\infty$ the average of the outputs tends to the exact answer. This is a
+MC algorithm.
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:481]: I'm standardizing to `Monte Carlo`. Thank you.

File: books/RayTracingTheRestOfYourLife.html [hunk 16]
```diff
@@ -423,184 +571,504 @@
...
-height of all the bins would drop to zero. A density function is one where we take the bins and
-adjust them so they don’t get shorter as we add more bins. For the case of the tree histogram above
-we might try:
+If we had more trees in our data source, the number of bins would stay the same, but each bin would
+have a higher frequency of tree. If we divided the data into more bins, we'd have more bins, but
+each bin would have a lower frequency of tree. If we took the number of bins and raised it to
+infinity, we'd have an infinite number of zero-frequency bins. To solve for this, we'll replace our
+histogram, which is a _discrete function_, with a _discrete density function_. A
+_discrete density function_ differs from a _discrete function_ in that it normalizes the y-axis to a
+fraction or percentage of the total, _i.e_ it's density, instead of a total count for each bin.
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:658]: I seem to do this alot

File: books/RayTracingTheRestOfYourLife.html [hunk 17]
```diff
@@ -423,184 +571,504 @@
...
 
 <div class='together'>
-If we added data for more trees, the histogram would get taller. If we divided the data into more
-bins, it would get shorter. A discrete density function differs from a histogram in that it
-normalizes the frequency y-axis to a fraction or percentage (just a fraction times 100). A
-continuous histogram, where we take the number of bins to infinity, can’t be a fraction because the
-height of all the bins would drop to zero. A density function is one where we take the bins and
-adjust them so they don’t get shorter as we add more bins. For the case of the tree histogram above
-we might try:
+If we had more trees in our data source, the number of bins would stay the same, but each bin would
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:652]: Yeah. This histogram section is a bit confusing. The initial wording had heights of trees, heights of bins.
I was tempted to just rip it out and rewrite this whole section, but I just don't have time for that.

I'm go through and try to generalize `tree` to `item` where it makes sense.

Thank you.

File: books/RayTracingTheRestOfYourLife.html [hunk 18]
```diff
@@ -423,184 +571,504 @@
...
-able to find some function $f(d)$ that gives us what we want. Suppose $e = f(d) = d^2$. This is no
-longer a uniform PDF. The PDF of $e$ will be bigger near 1 than it is near 0 (squaring a number
-between 0 and 1 makes it smaller). To convert this general observation to a function, we need the
-cumulative probability distribution function $P(x)$:
+After spending enough time with PDFs you might start referring to a PDF as the probability that a
+variable $r$ is value $x$, _i.e._ $p(r=x)$. Don't do this. For a continuous function, The
+probability that a variable is a specific value is always zero. If you ask "given the PDF $p(r)$,
+what are the odds that $r=1$"? For all PDFs this value is always zero. A PDF can only tell you the
+probability that a variable will fall within an interval. If the interval you're checking against is
+a constant number, then the PDF will always return a zero probability.
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:741]: Okee Doke

File: books/RayTracingTheRestOfYourLife.html [hunk 19]
```diff
@@ -423,184 +571,504 @@
...
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+That's an easy case, but the vast majority of cases that we're going to care about are nonuniform.
+We need to figure out a way to convert a uniform random number generator into a nonuniform random
+number generator, where the distribution is defined by the PDF. We'll just _assume_ that there
+exists a function $f(d)$ that takes uniform input and produces a nonuniform distribution weighted by
+PDF. Now we need to figure out a way to solve for $f(d)$.
+
+For the PDF given above, where $p(r) = \frac{r}{2}$, the probability of a random sample is higher
+toward 2 than it is toward 0. More 1.8s and 1.9s should be generated than 0.1s and 0.2s. If we put
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:786]: If I say that the numberd 1.8 and 1.9 have a higher chance than 0.1 and 0.2, then I am saying that a specific number has a discrete probability of happening. This contradicts my hammering of the point above that the probability of any number being generated is zero.

I think I'll just rewrite something like "greater probability of getting a number between 1.8 and 2.0 than 0.0 and 0.2"

[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:786]: Yes, I know that _our_ RNG is discrete, but it would muddy the text to mention that point in the text...

File: books/RayTracingTheRestOfYourLife.html (hunk 12 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:1024]: It was in the original text. I'll noodle of keeping/tossing

File: books/RayTracingTheRestOfYourLife.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:786]: Actually, I was just rewording what you said, but I see what you mean about specific values. I agree that it might be an improvement to talk about one range having more probability than some other range.

File: books/RayTracingTheRestOfYourLife.html (hunk 13 above)
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:1182]: Oops. “Those were most of the concepts ...”

File: src/InOneWeekend/main.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (hollasch) [reply src/InOneWeekend/main.cc:78]: I'll let you decide. I mostly don't like the anachronism of int-returning but never actually returning, as a common compiler backwards-compatible sleight-of-hand.

You choose:
1. `void main()`, no explicit return.
2. `int main()`, return 0.

Either works for me, and from your input, `int main()` may cause fewer warnings for readers of all platforms.

File: books/RayTracingTheRestOfYourLife.html (hunk 11 above)
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:974]: Which reminds me of a dad joke I just ran across. I wanted to become a Gregorian monk, but I never got the chants.

**Review by hollasch — APPROVED** [review]
Missing some `return 0;`s, but so far so good.

**Comment by trevordblack** [comment]
It felt gratuitous to add `return 0` to literally all of the `main` functions. Is that what you want? Cause I find it unnecessary and kind of ugly.

File: books/RayTracingTheRestOfYourLife.html (hunk 11 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:975]: At this rate, none of us will get the chants.

**Comment by hollasch** [comment]
If it's a hassle, don't worry about it. Just bugs me that if your function is named m-a-i-n, then you declare a return type but don't actually have to fulfill it, unlike every other function. But I know I'm just being crotchety. Go ahead and undo my changes, leave them as int-returning, without return statements. They should at least be consistent.

File: books/RayTracingTheRestOfYourLife.html [hunk 20]
```diff
@@ -736,10 +739,10 @@
 <div class='together'>
 After spending enough time with PDFs you might start referring to a PDF as the probability that a
 variable $r$ is value $x$, _i.e._ $p(r=x)$. Don't do this. For a continuous function, The
-probability that a variable is a specific value is always zero. If you ask "given the PDF $p(r)$,
-what are the odds that $r=1$"? For all PDFs this value is always zero. A PDF can only tell you the
+probability that a variable is a specific value is always zero. A PDF can only tell you the
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:742]: "The" is mistakenly capitalized.

Also, what do you think about this? “the probability of any single value is always zero, because its "bin" is infinitely thin (has zero width).”

File: books/RayTracingTheRestOfYourLife.html [hunk 21]
```diff
@@ -736,10 +739,10 @@
...
 probability that a variable will fall within an interval. If the interval you're checking against is
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:743]: Suggestion: “... probability that a variable will fall within a given interval of non-zero width.”

File: books/RayTracingTheRestOfYourLife.html [hunk 22]
```diff
@@ -736,10 +739,10 @@
...
-a constant number, then the PDF will always return a zero probability.
+a constant number, then the PDF will always return a zero probability. Here's a simple mathematical
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:744]: “... is a single value, then the PDF ...”

File: books/RayTracingTheRestOfYourLife.html [hunk 23]
```diff
@@ -784,12 +787,12 @@
...
 
 For the PDF given above, where $p(r) = \frac{r}{2}$, the probability of a random sample is higher
-toward 2 than it is toward 0. More 1.8s and 1.9s should be generated than 0.1s and 0.2s. If we put
-aside our mathematics hat for a second and put on our computer science hat, maybe we can figure out
-a smart way of partitioning the PDF. We know that there is a higher probability near 2 than near
-0, but what is the value that splits the probability in half? What is the value that a number chosen
-at random has a 50% chance of being higher than and 50% chance of being lower than? What is the $x$
-that solves:
+toward 2 than it is toward 0. There is a greater probability of getting a number betwween 1.8 and
+2 than 0 and 0.2. If we put aside our mathematics hat for a second and put on our computer
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:791]: “... between [note typo] 1.8 and 2.0 than there is between 0.0 and 0.2.”

File: books/RayTracingTheRestOfYourLife.html [hunk 24]
```diff
@@ -829,9 +832,9 @@
 
     ![Figure [approx-f]: A nonuniform distribution for r/2](../images/fig-3.06-nonuniform-dist.jpg)
 
-We had the analytical solution to the integration above, so we could very very easily solve for the
-50% value. We could also solve for this experimentally. If we had a function that couldn't solve
-for the integration for--or didn't _want_ to solve the integration for--we could get an experimental
+We had the analytical solution to the integration above, so we could very easily solve for the 50%
+value. But we could also solve for this 50% value experimentally. There will be functions that we
+either can't or don't want to solve the integreation for. In these cases we can get an experimental
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:837]: *integration

“to solve for the integration”

"In these cases, we can ...”

File: books/RayTracingTheRestOfYourLife.html [hunk 25]
```diff
@@ -840,7 +843,12 @@
 
     ![Figure [exp-sin2]: A function that we don't want to solve analytically](../images/fig-3.07-exp-sin2.jpg)
 
-From $0$ to $2\pi$ as example:
+At this point you should be familiar with how to experimentally solve for the area under a curve.
+We'll be taking our existing code and modifying it slightly to get an estimate for the 50% value. We
+want to solve for the $x$ value that gives us half of the total area under the curve. As we go along
+and solve for the sum over `N` samples, we're also going to store these `N` samples and their
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:849]: I'd drop the backticks around N.

File: books/RayTracingTheRestOfYourLife.html [hunk 26]
```diff
@@ -840,7 +843,12 @@
...
+associated `p(x)` value. After we solve for the total sum we'll go back through our samples once
+more and count up until we find an area that is half of the total. From $0$ to $2\pi$ as example:
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:851]: “as example” → “for example”

File: books/RayTracingTheRestOfYourLife.html [hunk 27]
```diff
@@ -930,21 +946,20 @@
...
 ---------------------------
 
-This was a lot of math and work to build up a couple of notions. Let's return to our initial PDF and
-enumerate a new term. Where we don't explicitly set a probability, we assume to be zero. So for our
-example from the beginning of the chapter, $p(r) = 0$, for $r < 0$ and $r > 2$. _i.e._, the
-probability of generating a random number there is zero. We can rewrite our $p(r)$ in piecewise
-fashion:
+This was a lot of math and work to build up a couple of notions. Let's return to our initial PDF. We
+assume the PDF to be zero where we don't explicitly set a probability. So for our example from the
+beginning of the chapter, $p(r) = 0$, for $r < 0$ and $r > 2$. That is, the probability of
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:951]: Maybe tweak to “..., $p(r) = 0$ for $r \notin [0,2]$”

**Review by hollasch — APPROVED** [review]

File: books/RayTracingTheRestOfYourLife.html [hunk 28]
```diff
@@ -624,8 +627,8 @@
 a surface, one that is randomly reflected toward the light and one that is not. The ray that is
 reflected toward the light will appear a very bright color. The ray that is reflected to somewhere
 else will appear a very dark color. The two intensities should really be somewhere in the middle.
-We could lessen this problem if we sent _more_ random samples toward this light, but this will cause
-the scene to be inaccurately bright.
+We could lessen this problem if we biased both of these rays so that they steered toward toward the
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:630]: EMBARRASSING.

File: books/RayTracingTheRestOfYourLife.html (hunk 23 above)
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:791]: That was deliberate, but I'll fix it.

**Comment by hollasch** [comment]
Looks good!
